src/lib/lightning/clightning/clightningService.ts

```diff
@@ -197,9 +199,100 @@ class CLightningService implements LightningService {
     );
   }
 
+  async addListenerToNode(node: LightningNode): Promise<void> {
+    console.log('addListenerToNode CLN on port: ', node.ports.rest);
+  }
+
+  private channelCaches: {
+    [nodePort: number]: {
+      intervalId: NodeJS.Timeout;
+      channels: { pending: boolean; status: string }[];
+    };
+  } = {};
+
+  async removeListener(node: LightningNode): Promise<void> {
+    const nodePort = this.getNodePort(node);
+    const cache = this.channelCaches[nodePort];
+    if (cache) {
+      clearInterval(cache.intervalId);
+      delete this.channelCaches[nodePort];
+    }
+  }
+
+  async subscribeChannelEvents(
+    node: LightningNode,
+    callback: (event: PLN.LightningNodeChannelEvent) => void,
+  ): Promise<void> {
+    const nodePort = this.getNodePort(node);
+    if (!this.channelCaches[nodePort]) {
+      // check c-lightning channels every 30 seconds
+      this.channelCaches[nodePort] = {
+        intervalId: setInterval(() => {
+          this.checkChannels(node, callback);
+        }, 30 * 1000),
+        channels: [],
+      };
+    }
+  }
+
+  checkChannels = async (
+    node: LightningNode,
+    callback: (event: PLN.LightningNodeChannelEvent) => void,
+  ) => {
+    const result = await httpGet<CLN.GetChannelsResponse[]>(node, 'channel/listChannels');
+    const channels = result.map(c => {
+      const status = ChannelStateToStatus[c.state];
+      return { pending: status !== 'Open', status };
+    });
+
+    const cache = this.channelCaches[this.getNodePort(node)];
+
+    if (!this.areChannelsEqual(cache.channels, channels)) {
+      this.handleUpdatedChannels(channels, callback);
+      cache.channels = channels;
+      // edge case for empty channels but cache channels is not empty
+      if (!channels.length) {
+        callback({ type: 'Closed' });
+      }
+    }
+  };
+
+  private handleUpdatedChannels = (
+    channels: { pending: boolean; status: string }[],
+    callback: (event: PLN.LightningNodeChannelEvent) => void,
+  ) => {
+    channels.forEach(channel => {
+      if (channel.pending) {
+        callback({ type: 'Pending' });
+      } else if (channel.status === 'Open') {
+        callback({ type: 'Open' });
+      } else if (channel.status === 'Closed') {
+        callback({ type: 'Closed' });
+      }
+    });
+  };
+
   private toSats(msats: number): string {
     return (msats / 1000).toFixed(0).toString();
   }
+
+  private areChannelsEqual(
+    channels1: { pending: boolean; status: string }[],
+    channels2: { pending: boolean; status: string }[],
+  ): boolean {
+    return (
+      channels1.length === channels2.length &&
+      JSON.stringify(channels1) === JSON.stringify(channels2)
+    );
+  }
```

Comment:
This implementation could result in a false positive if the order of the channels are not the same in subsequent calls. It would be better to include the channelId in the cached data and compare each channel by ID. THis will also help resolve my comment above because you can selectively call the `callback` only for the channels that have changed. This will also need to handle cases where a channel has been added or removed.

Reply:
Updated

Reply:
Nice improvement. 👌